andela/wolfsbane-backend
# #167750059 Build out the Trip-Request endpoints [CRUD]

**What does this PR do?**
This PR holds the built out **CRUD** endpoints for [Travel-Requests]

**Description of tasks**
The following **endpoints** should be working;
- POST **/api/v1/request/:departmentId** _[for users to create a request]_
- DELETE **/api/v1/request/:id** _[for users to delete a request, when it is still pending]_
- GET **/api/v1/request/:id** _[for users to get a request]_
- GET **/api/v1/request** _[for users to get all requests]_
- POST **/api/v1/trip/:requestId** _[for users to create a trip under a request]_
- PUT **/api/v1/trip/:requestId/:tripId** _[for users to update a trip only when it is pending]_
- DELETE **/api/v1/request/:id** _[for users to delete a request, when it is still pending]_
- GET **/api/v1/request/manager** _[for a manager to get all requests,made by any member of his department only]_

**_after_**

- building out the controllers, middle-wares and services
- creating routes to access the endpoints
- writing tests for each of the endpoints

**How should this be manually tested/checked?**
_let's get started_
- Clone this PR using `git fetch origin pull/24/head:ft-make-trip-request-167750058`
- run npm install
- Drop your databases for the test & dev environment and run **_npm pre-dev_** and **_npm pre-test_**
- Start your server.
- Now go to postman and place in this route: **_localhost:3000/api/v1/users/signin_**
- send this **POST** request: 
```
{
    "email" : "kelechi@gmail.com",
    "password" : "kelechi"
}
```
-  copy the token generated, **_have it handy_**

TO CREATE A REQUEST
- _**look up to get a proper idea of the api endpoints**_
- place in this route :
- `localhost:3000/api/v1/request/36da1ce3-8e1f-4b0f-8e15-1189d8231ef2` on postman
`NOTE: 36da1ce3-8e1f-4b0f-8e15-1189d8231ef2 is a dept. ID which is already in our db when we seeded.`
- send this **POST** request of the line manager mail:
```
{
	"lineManagerMail": "<whatever-mail>"
}
```
- copy the token generated from sign in and place it in your Bearer Token Authorization tab
- then hit send, a response should be generated.

TO CREATE A TRIP
- MAKE sure the token you used in signing in and creating a request is **maintained**
- open another tab on postman and place in this route- `localhost:3000/api/v1/trip/<request-id>NOTE: the ID of the request you created.`
- send this **POST** request
```
{
	"origin": "lagos",
	"destination": "nairobi",
	"departureDate": "2019-09-09 01:33:28.862+00",
	 "returnDate": "2019-09-09 01:33:28.862+00",
	"travelReasons": "To get away.",
	"typeOfTrip": "Return",
	"roomId": "5402e279-9447-4ac6-98c8-44cf10e27f4e",
	"accommodationId": "777f640e-a2ff-45ee-9ce1-bf37645c42d6"
}

```
Now, TO UPDATE THE TRIP YOU CREATED, **_when it is still pending_**
- Maintain your token please
- Open another tab on postman and place in this route -  `localhost:3000/api/v1/trip/<request-id>/<trip-id>` **both the ids we got when we created a request and a trip**
- Now send this **PUT** request
```
{
	"origin": "lagos",
	"destination": "uganda",
	"departureDate": "2019-09-09 01:33:28.862+00",
	"travelReasons": "To get away.",
	"typeOfTrip": "One Way",
	"roomId": "5402e279-9447-4ac6-98c8-44cf10e27f4e",
	"accommodationId": "777f640e-a2ff-45ee-9ce1-bf37645c42d6"
}

```
NOW TO GET THE REQUEST YOU CREATED, _since we have created a trip with the request ticket and even updated it._
- Maintain your token
- Open another tab on postman and place in this route - `localhost:3000/api/v1/request<request-id>` (the id of the request that was created)

NOW TO GET ALL THE REQUESTS YOU HAVE CREATED, _which is actually just one yet._
- Maintain your token
- Open another tab on postman and place in this route - `localhost:3000/api/v1/request` 

NOW, TO DELETE A REQUEST(TRIP) that is still **_pending_**
- Maintain your token please
- Open another tab on postman and place in this route-  `localhost:3000/api/v1/<request-id>` **(the id of the request we created**

NOW, FOR A MANAGER TO GET ALL REQUESTS
- open another tab on postman and place in this route:  `localhost:3000/api/v1/users/signin`
- send this **POST** request to log in a manager
```
{
    "email" : "adelekegbolahan92@yahoo.com",
    "password" : "funmi1234"
}
``` 
- Now, get the token generated and open another tab on postman, add the token to the Bearer Token field, and place in this route -` localhost:3000/api/v1/request/manager`
- send a GET request with the route above to get the information
**To query the status of any request state as a manager:**
**_for example:_**
**GET**
` localhost:3000/api/v1/request/manager?status=pending`

---------------------------------------------------------------------------------------------------------------
**Let us do some security checks**, _shall we?_
1. 
_**for manager getting requests**_
- _Adeleke Gbolahan is a manager and kelechi is not._
- send this POST request using the route ` localhost:3000/api/v1/users/signin`
```
{
    "email" : "kelechi@gmail.com",
    "password" : "kelechi"
}
```
- get the token and try to access this route: ` localhost:3000/api/v1/request/manager`
- _you should get a response denying your access._
- _But anybody logged in that is a super admin will_ 
---------------------------------------------------------------------------------------------------------------
2. 
Trying to delete a trip-request that is still pending
- sign in this user:
```
{
    "email" : "samailabalap@gmail.com",
    "password" : "funmi1234"
}
```
- get the token from the sign in and place it in the Bearer Token tab, also place this route on postman- ` localhost:3000/api/v1/delete` and try to delete
- the response will deny you an access.
-------------------------------------------------------------------------------------------------------------
**Other security checks you can do:**
- a user trying to update a trip when the status is not pending
- a user trying to get a/all request(s) that she did not create
- a user trying to create a request with a department that does not exist.
-------------------------------------------------------------------------------------------------------------

**Any background context you want to provide?**
- Make sure to drop both dbs and run _npm run pre-dev and npm run pre-test_
- few changes were made to the models; 
- **_typeOfTrip_** was added to the trip models
- the _**accommodationId and association**_ that sat on the Requests table was moved and associated to the Trips (so the definition makes more sense)
- an email field was added to the Requests table for a requester to include the line manager's mail for notifications

**What are the relevant pivotal tracker stories?**
<a href="https://www.pivotaltracker.com/story/show/167750058">#167750058</a>
